File: sdks/java/basys.sdk/src/main/java/org/eclipse/basyx/submodel/metamodel/map/qualifier/haskind/HasKind.java
package org.eclipse.basyx.submodel.metamodel.map.qualifier.haskind;

import java.util.Map;

import org.eclipse.basyx.submodel.metamodel.api.qualifier.haskind.IHasKind;
import org.eclipse.basyx.submodel.metamodel.api.qualifier.haskind.ModelingKind;
import org.eclipse.basyx.vab.model.VABModelMap;

/**
 * HasKind class
 * 
 * @author elsheikh, schnicke
 *
 */
public class HasKind extends VABModelMap<Object> implements IHasKind {
	public static final String KIND = "kind";

	/**
	 * Constructor
	 */
	public HasKind() {
		// Default value

		put(KIND, null);
	}

	/**
	 * Constructor that takes
	 * {@link org.eclipse.basyx.submodel.metamodel.map.qualifier.haskind.Kind
	 * Kind}(either Kind.Instance or Kind.Type)
	 */
	public HasKind(ModelingKind kind) {
		// Kind of the element: either type or instance.

		put(KIND, kind.toString());
	}

	/**
	 * Creates a HasKind object from a map
	 * 
	 * @param obj
	 *            a HasKind object as raw map
	 * @return a HasKind object, that behaves like a facade for the given map
	 */
	public static HasKind createAsFacade(Map<String, Object> map) {
		if (map == null) {
			return null;
		}

		HasKind ret = new HasKind();
		ret.setMap(map);
		return ret;
	}

	@Override
	public ModelingKind getModelingKind() {
		String str = (String) get(HasKind.KIND);
		return ModelingKind.fromString(str);
	}

	public void setModelingKind(ModelingKind kind) {
		put(HasKind.KIND, kind.toString());
	}

}

File: sdks/c++/basys.sdk.cc/src/submodel/submodel/map/qualifier/HasKind.cpp
/*
 * HasKind.cpp
 *
 *      Author: wendel
 */

#include <BaSyx/submodel/map/qualifier/HasKind.h>

using namespace basyx::submodel;

HasKind::HasKind(Kind kind) 
	: vab::ElementMap()
{
	this->Init(kind);
}

void HasKind::Init(Kind kind)
{
	this->map.insertKey(Path::Kind, util::to_string(kind));
};

HasKind::HasKind(basyx::object object)
	: vab::ElementMap(object)
{
}

basyx::submodel::HasKind::HasKind(const IHasKind & other) :
  vab::ElementMap{}
{
  this->setHasKindReference(other.getHasKindReference());
}

Kind HasKind::getHasKindReference() const
{
	return util::from_string<Kind>(this->map.getProperty(Path::Kind).GetStringContent());
}

void HasKind::setHasKindReference(Kind kind)
{
	this->map.insertKey(Path::Kind, util::to_string(kind), true);
}

static const std::string kind_to_string[] = {
	"NotSpecified",
	"Type",
	"Instance"
};

static const std::unordered_map<std::string, Kind> string_to_kind{
	{ kind_to_string[static_cast<char>(Kind::Instance)] , Kind::Instance },
	{ kind_to_string[static_cast<char>(Kind::Type)] , Kind::Type },
	{ kind_to_string[static_cast<char>(Kind::NotSpecified)] , Kind::NotSpecified }
};

const std::string & util::to_string(Kind kind)
{
	return kind_to_string[static_cast<char>(kind)];
}

template<>
Kind util::from_string<Kind>(const std::string & str)
{
	return string_to_kind.at(str);
};
